### PR TITLE
Fix error thrown when inserting keys

### LINKED ISSUE #69: Duplicate mapping key and parsing error when trying to add a map entry
I'm trying to add a key to a map and set its value, but running into issues.

```yaml
# Input Yaml
- name: foo
  id: "A"
- name: bar

# Desired output Yaml
- name: foo
  id: "A"
- name: bar
  id: "B"
```

Example code for Duplicate mapping key error
```dart
import 'package:yaml_edit/yaml_edit.dart';

void main() {
  const yaml = '''
- name: foo
  id: "A"
- name: bar
  ''';
  final yamlEditor = YamlEditor(yaml);
  yamlEditor.update(
    [1, 'id' ],
    "B",
  );
  print(yamlEditor);
}
```

For this, I get the following error:
```bash
Unhandled exception:
Error on line 3, column 3: Duplicate mapping key.
  ╷
3 │   id: B
  │   ^^
  ╵
#0      Loader._loadMapping (package:yaml/src/loader.dart:171:9)
#1      Loader._loadNode (package:yaml/src/loader.dart:92:16)
#2      Loader._loadSequence (package:yaml/src/loader.dart:146:20)
#3      Loader._loadNode (package:yaml/src/loader.dart:90:16)
#4      Loader._loadDocument (package:yaml/src/loader.dart:68:20)
#5      Loader.load (package:yaml/src/loader.dart:60:20)
#6      loadYamlDocument (package:yaml/yaml.dart:72:25)
#7      loadYamlNode (package:yaml/yaml.dart:57:5)
#8      YamlEditor._performEdit (package:yaml_edit/src/editor.dart:576:24)
#9      YamlEditor.update (package:yaml_edit/src/editor.dart:271:14)
#10     main (package:yaml_playground/example.dart:11:14)
#11     _delayEntrypointInvocation.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:297:19)
#12     _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:184:12)
```

If I swap the order around and try to update the first list item instead, I get a different error
```dart
import 'package:yaml_edit/yaml_edit.dart';

void main() {
  const yaml = '''
- name: foo
- name: bar
  id: "B"
  ''';
  final yamlEditor = YamlEditor(yaml);
  yamlEditor.update(
    [0, 'id' ],
    "A",
  );
  print(yamlEditor);
}
```
```bash
Unhandled exception:
Error on line 2, column 1: Only expected one document.
  ╷
2 │ ┌ - name: foo
3 │ │ - name: bar
4 │ │   id: "B"
5 │ └   
  ╵
#0      loadYamlDocument (package:yaml/yaml.dart:80:5)
#1      loadYamlNode (package:yaml/yaml.dart:57:5)
#2      YamlEditor._performEdit (package:yaml_edit/src/editor.dart:576:24)
#3      YamlEditor.update (package:yaml_edit/src/editor.dart:271:14)
#4      main (package:yaml_playground/example.dart:11:14)
#5      _delayEntrypointInvocation.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:297:19)
#6      _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:184:12)
```

In a larger example that I can't duplicate in this smaller example, the error I get in this case is: `Expected a key while parsing a block mapping.`

Using version yaml_edit: ^2.2.0

### PR DESCRIPTION
**Pull request by kekavc24**
Closes #69 

- Checks for ordering only when more than one key is present
---

- [x] I’ve reviewed the contributor guide and applied the relevant portions to this PR.

<details>
  <summary>Contribution guidelines:</summary><br>

- See our [contributor guide](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md) for general expectations for PRs.
- Larger or significant changes should be discussed in an issue before creating a PR.
- Contributions to our repos should follow the [Dart style guide](https://dart.dev/guides/language/effective-dart) and use `dart format`.
- Most changes should add an entry to the changelog and may need to [rev the pubspec package version](https://github.com/dart-lang/sdk/wiki/External-Package-Maintenance#making-a-change).
- Changes to packages require [corresponding tests](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md#Testing).

Note that many Dart repos have a weekly cadence for reviewing PRs - please allow for some latency before initial review feedback.
</details>

### CODE REVIEW

**Comment by kekavc24** [comment]
@jonasfj Before I can add new test files. It seems many of the old tests were based on this bug  e.g.

https://github.com/dart-lang/yaml_edit/blob/31919348bd2a1bbb805b4eb88a6b7f50d4ab247e/test/editor_test.dart#L23-L35

The fix now returns the new correct `SourceEdit` as the correct offset for the edit `32` not `0` anymore.

**Comment by jonasfj** [comment]
By all means fix all the old tests too.

But are you sure wise to support reverse alphabetical ordering?
It might be unlikely that someone will do that intentionally.

Especially if there is only 2 properties then reverse alphabetical ordering is really just any order that isn't alphabetical :)

**Comment by kekavc24** [comment]
Oh, yeah. Sorry for the overkill. I'll fix that and the tests. Thanks.

**Comment by kekavc24** [comment]
@jonasfj I'm still debugging the `spliceList` method. It seems something broke somewhere.

File: lib/src/utils.dart [hunk 1]
```diff
@@ -124,16 +124,21 @@ bool isFlowYamlCollectionNode(Object value) =>
 int getMapInsertionIndex(YamlMap map, Object newKey) {
   final keys = map.nodes.keys.map((k) => k.toString()).toList();
 
-  for (var i = 1; i < keys.length; i++) {
-    if (keys[i].compareTo(keys[i - 1]) < 0) {
-      return map.length;
+  final numOfKeys = keys.length;
+
+  // We can't deduce ordering if list is empty or one key is present
+  if (numOfKeys > 1) {
```
Comment (jonasfj) [inline lib/src/utils.dart:130]: This is fine, but sometimes life becomes easier by returning early:
```dart
if (numOfKeys <= 1) {
  return map.length;
}
```

File: test/testdata/output/allowed_characters_in_keys.golden
Comment (jonasfj) [inline test/testdata/output/allowed_characters_in_keys.golden:1]: This is happening because the ordering is changing, because this test starts with a map of size 1 and then just adds more keys. Which previously would be ordered, and now won't be.

hmm, I could see a point to ordering maps that have only 1 key. Since we apparently have a lot of tests doing exactly this.

@kekavc24 What do you think we should do? Should we order keys in a map when there is only 1 key?

I don't mind landing this as is.

File: test/update_test.dart [hunk 2]
```diff
@@ -876,6 +881,46 @@ d: 4
...
+        doc.update(['d'], 4);
+        expect(doc.toString(), equals('''
+a: 1
+c: 2
+b: 3
+d: 4
+
+
+'''));
+      });
```
Comment (jonasfj) [inline test/update_test.dart:923]: These tests are awesome, have you considered add some tests in `test/testdata/input/`?

They are less powerful than this kind of testing, but it's much easier to write a lot of test cases and therefore easier to make sure that more cases are covered.

**Review by jonasfj — COMMENTED** [review]
Other than tests still failing, I think this pretty good.

I could be convinced that we should sort new entries in maps that only contain a single key.
Given that we already have tests that rely on this behavior.

One could also argue that we should allow ordering to be configurable in `YamlEditor` -- but that's probably overkill. I wouldn't mind doing it, but if we want to pursue such things  I'd suggest we do it in a separate PR.

**Comment by jonasfj** [comment]
By the way, the test that is failing is somewhere in `test/random_test.dart`.

I'd suggest trying to add more print statements and extract the YAML + the modification being made into a separate test.

Don't try to debug things in `random_test.dart`, try to extract the test case it generates instead. We'll want it as a hardcoded test case anyways :D

**Comment by jonasfj** [comment]
I noticed a few weird hacks with prints in `random_test.dart` and filed: https://github.com/dart-lang/yaml_edit/pull/81 

After that the failing test case should print:
```dart
Failed to call YamlModificationMethod.splice on:

environment: true


0.03167456183131434:
  - - true  
    - |-
        false
    - 4177511653
  - null
  - 2910742758
  - true
  - - false 
  - 2081477285
  - 1718060652
  - 0.9662696464087974
  - null
  - "9[hd24=s438Y7a8_!M&}s<3cTr{\"e%<1I7vq&xp4p\"9jH#NJv"
762977104: null

with the following arguments:
[0, 1, [0.7396543051030952, null, {P/wQ'@G`$A%6LA$aXk0Uy&f36Pi[~:905@`DJ~nKUs!2:$45hj[vk7V<Qt~\2M^Uk$, vZEjAW=tOT=Q&dpH,cN'hX$_wHR=:q?UvHzC;+8*bLX_5Rgvp)L{E, {null: null, {3368549847: null, 0.9104247116161583: dcN&=EiSL?{sWLMRg -W]%
[=~Ggy@Z9F<77:3dY!{wE;m\uC9R9S2sIv<LyIUK, JRBV\|1]O<K;Wosal33 f30@`?*{}?}ke>^Eg''oUXw+JXq%C|{Ck._BGk: true}: ヽ༼ຈل͜ຈ༽ﾉ ヽ༼ຈل͜ຈ༽ﾉ}, 1853891542]]
and path:   
[0.03167456183131434, 4]

Error Details:
Assertion failed: (package:yaml_edit) Modification did not result in expected result.

# YAML before edit:
>
> environment: true
>
>
> 0.03167456183131434:
>   - - true
>     - |-  
>         false
>     - 4177511653
>   - null  
>   - 2910742758
>   - true  
>   - - false
>   - 2081477285
>   - 1718060652
>   - 0.9662696464087974
>   - null
>   - "9[hd24=s438Y7a8_!M&}s<3cTr{\"e%<1I7vq&xp4p\"9jH#NJv"
> 762977104: null
>

# YAML after edit:
>
> environment: true
>
>
> 0.03167456183131434:
>   - - true
>     - |-
>         false
>     - 4177511653
>   - null
>   - 2910742758
>   - true
>     - 1853891542
>   - - false
>   - 2081477285
>   - 1718060652
>   - 0.9662696464087974
>   - null
>   - "9[hd24=s438Y7a8_!M&}s<3cTr{\"e%<1I7vq&xp4p\"9jH#NJv"
> 762977104: null
>

Please file an issue at:
https://github.com/dart-lang/yaml_edit/issues/new?labels=bug


#0      YamlEditor._performEdit (package:yaml_edit/src/editor.dart:579:7)
#1      YamlEditor.insertIntoList (package:yaml_edit/src/editor.dart:339:5)
#2      YamlEditor.spliceList (package:yaml_edit/src/editor.dart:376:7)
#3      _Generator.performNextModification (file:///workspace/yaml_edit/test/random_test.dart:209:20)
#4      main.<anonymous closure>.<anonymous closure> (file:///workspace/yaml_edit/test/random_test.dart:47:27)
#5      _ReturnsNormally.typedMatches (package:matcher/src/core_matchers.dart:153:8)
#6      FeatureMatcher.matches (package:matcher/src/feature_matcher.dart:16:42)
#7      _expect (package:matcher/src/expect/expect.dart:138:30)
#8      expect (package:matcher/src/expect/expect.dart:56:3)
#9      main.<anonymous closure> (file:///workspace/yaml_edit/test/random_test.dart:46:9)
#10     Declarer.test.<anonymous closure>.<anonymous closure> (package:test_api/src/backend/declarer.dart:215:19)
<asynchronous suspension>
#11     Declarer.test.<anonymous closure> (package:test_api/src/backend/declarer.dart:213:7)
<asynchronous suspension>
#12     Invoker._waitForOutstandingCallbacks.<anonymous closure> (package:test_api/src/backend/invoker.dart:258:9)
<asynchronous suspension>


00:03 +15 -1: testing with randomly generated modifications: test 15 [E]
  Expected: return normally
    Actual: <Closure: () => void>
     Which: threw _YamlAssertionError:<Assertion failed: (package:yaml_edit) Modification did not result in expected result.

            # YAML before edit:
            >
            > environment: true
            >
            >
            > 0.03167456183131434:
            >   - - true
            >     - |-
            >         false
            >     - 4177511653
            >   - null
            >   - 2910742758
            >   - true
            >   - - false
            >   - 2081477285
            >   - 1718060652
            >   - 0.9662696464087974
            >   - null
            >   - "9[hd24=s438Y7a8_!M&}s<3cTr{\"e%<1I7vq&xp4p\"9jH#NJv"
            > 762977104: null
            >

            # YAML after edit:
            >
            > environment: true
            >
            >
            > 0.03167456183131434:
            >   - - true
            >     - |-
            >         false
            >     - 4177511653
            >   - null
            >   - 2910742758
            >   - true
            >     - 1853891542
            >   - - false
            >   - 2081477285
            >   - 1718060652
            >   - 0.9662696464087974
            >   - null
            >   - "9[hd24=s438Y7a8_!M&}s<3cTr{\"e%<1I7vq&xp4p\"9jH#NJv"
            > 762977104: null
            >

            Please file an issue at:
            https://github.com/dart-lang/yaml_edit/issues/new?labels=bug
            >

  package:matcher             expect
  test/random_test.dart 46:9  main.<fn>
```

And from that we should be able to extract the test case.
Once extracted we can craft a minimal example from the test case.

I think it makes sense to have the failing test cases and minimum example as hardcoded test cases going forward.

`random_test.dart` does make consistent test cases, but the RNG might change in the future and then it'll test other things.

----------

Side note: it's pretty cool that `random_test.dart` actually found an issue. It was really a shot in the dark when we made it :rofl: I had no idea that it would actually prove useful.

(Okay, we don't know that the issue it found is a useful error, but it hopefully is something worth fixing).

**Comment by jonasfj** [comment]
@kekavc24 I landed https://github.com/dart-lang/yaml_edit/pull/81 I'd suggest rebasing this branch on `main`, then it'll be much easier to figure out what's wrong in `random_test.dart`.

**Comment by jonasfj** [comment]
Also please do consider adding an entry to `CHANGELOG.md`.
Once we've landed this, I think we should cut a release, it's a pretty important bugfix!

I filed https://github.com/dart-lang/yaml_edit/pull/82 to start the `CHANGELOG.md`.

**Comment by jonasfj** [comment]
hmm, I'm not sure I entirely follow.

But I have a few questions:
 * Is this only an issue given the bug fix we have in this PR?
   * I'm guessing this bug has been there a while, but that this PR just made it surface, right?
 * Are we only discovering the issue because of the bug fix we have in this PR?
   Because we no longer make 1 length maps sorted when inserting the second key.
   And this just happens to cause this bug to surface now.

I think if we can make test pass by sorting 1 length maps again, then we should do so.
AND then file an issue to fix the bug in `spliceList`. That way we can do it in a separate PR.

-------------------

If on the other hand, the bug fix we made here, cause this bug in `spliceList`, then we should fix it in this PR.

**Comment by jonasfj** [comment]
I'm pretty sure this is an existing bug that just surfaced, try the test case.
```
SLICE LIST IN NESTED BLOCK LIST
---
key:
- true
- - false
---
- [splice, [key, 1], 0, 1, ['test']]
```
Just:
 * Create `test/testdata/input/splicelist_in_nested_block_list.test` with the contents above.
 * Run `dart test test/golden_test.dart`
 
It fails both on `main` and on this PR.

---------

Question is why it only surfaces in `random_test.dart` in this PR.

**Comment by jonasfj** [comment]
If we switch the implementation to:
```dart
int getMapInsertionIndex(YamlMap map, Object newKey) {
  final keys = map.nodes.keys.map((k) => k.toString()).toList();

  // We can't deduce ordering if list is empty, so then we just we just append
  if (keys.isEmpty) {
    return map.length;
  }

  for (var i = 1; i < keys.length; i++) {
    if (keys[i].compareTo(keys[i - 1]) < 0) {
      return map.length;
    }
  }

  final insertionIndex =
      keys.indexWhere((key) => key.compareTo(newKey as String) > 0);

  if (insertionIndex != -1) return insertionIndex;

  return map.length;
}
```

Then we've still fixed the bug we set out to fix, right? @kekavc24 
We'll be inserting in sorted order, even when there is only 1 entry in the map, but that's also how it worked before.

And with this `random_test.dart` just happens to pass. It's deterministically random, but it certainly doesn't try all combinations.

--------------------

If we do that, then we've done the smallest possible change to fix the bug we're trying to fix. Then we can land this and move on the next bug. It's much easier if we don't let PRs grow too big.

I filed https://github.com/dart-lang/yaml_edit/issues/83 for the issue with `spliceList`. And it'd be great if we can fix that too -- just easier to do in a separate PR.

**Comment by kekavc24** [comment]
It's not from this fix.

File: test/update_test.dart (hunk 2 above)
Comment (kekavc24) [reply test/update_test.dart:923]: Should I move them?

File: test/testdata/output/allowed_characters_in_keys.golden
Comment (kekavc24) [reply test/testdata/output/allowed_characters_in_keys.golden:1]: I think we should keep the changes. The tests now mirror expected behaviour rather than the bug.

File: test/update_test.dart (hunk 2 above)
Comment (jonasfj) [reply test/update_test.dart:923]: I'd rather suggest making more or different tests. Also fine not to.

File: test/testdata/output/allowed_characters_in_keys.golden
Comment (jonasfj) [reply test/testdata/output/allowed_characters_in_keys.golden:1]: But then we trigger the issue in `random_test.dart`, so we either have to add `skip: true` to `random_test.dart` so we can land this.

Or walk back the change with not sorting 1-key maps. Because then we don't trigger issues in `random_test.dart`.

We do want to fix the `spliceList` issue, but ideally in a different PR.

-------------

And I don't know what solution is best, building sorted maps when adding multiple keys to a map with a single key, feel equally good to just appending the keys.


-----------------


TL;DR: Let's get the tests passing, so we can land. Either fixing `spliceList`, walking back the 1-key map change, or skipping tests (and later fixing `spliceList`).

Comment (kekavc24) [reply test/testdata/output/allowed_characters_in_keys.golden:1]: Maybe skip? Once this is done, I'll start looking into that issue

**Comment by kekavc24** [comment]
@jonasfj Wow, I'm sorry. I really fumbled this. I committed a file I was meant to delete. 

My golden tests seem to be failing with no cause. Can you take a look?

**Comment by jonasfj** [comment]
Make sure you delete the output files before running golden tests... Then review the newly generated output.

Otherwise, I can look tomorrow maybe.

**Review by jonasfj — APPROVED** [review]

**Comment by kekavc24** [comment]
Currently on it. @jonasfj
